src/2025h2/autoreborrow-traits.md

```diff
@@ -0,0 +1,139 @@
+|------------------------------|---------------------|-------|
+| Discussion and moral support | ![Team][] [lang]    | Normal RFC process |
+| Help around impl             | ![Team][] [compiler]| Trait-impl querying in rustc to replace `Pin<&mut T>` special case |
+| Do the work                  | @aapoalas           |       |
+
+### Design autoreborrow internals
+
+The basic idea of autoreborrowing is simple enough: when a reborrowable type is encountered at a coercion
+site, attempt a reborrow operation.
+
+Complications arise when reborrowing becomes recursive: if a `struct X { a: A, b: B }` contains two
+reborrowable types `A` and `B`, then we'd want the reborrow of `X` to be performed "piecewise". As an
+example, the following type should, upon reborrow, only invalidate any values that depend on the `'a` lifetime while any values dependent on the `'b` lifetime should still be usable as normal.
+
+```rust
+struct X<'a, 'b> {
+    a: &'a mut A,
+    b: &'b B,
+}
+```
+
+To enable this, reborrowing needs to be defined as a recursive operation but what the "bottom-case" is, that
+is the question. One option would be to use `!Copy + Reborrow` fields, another would use core marker types
+like `PhantomExclusive<'a>` and `PhantomShared<'b>` to discern the difference.
+
+
+| Task                 | Owner(s) or team(s)                | Notes                                                               |
+|----------------------|------------------------------------|---------------------------------------------------------------------|
+| Lang-team experiment | ![Team][] [lang]                   | allows coding pre-RFC; only for trusted contributors                |
+| Author RFC           | *Goal point of contact, typically* |                                                                     |
+| Lang-team champion   | ![Team][] [lang]                   | Username here |
+| RFC decision         | ![Team][] [lang]                   |                                                                     |
+| Secondary RFC review | ![Team][] [types]                  | request bandwidth from a second team, most features don't need this |
+
+### Implement non-recursive autoreborrowing
+
+A basic autoreborrowing feature should not be too complicated: the `Pin<&mut T>` special-case in the
+compiler already exists and could probably be reimagined to rely on a `Reborrow` trait.
+
+| Task                              | Owner(s) or team(s)                | Notes |
+|-----------------------------------|------------------------------------|-------|
+| Implementation                    | *Goal point of contact, typically* |       |
+| Standard reviews                  | ![Team][] [compiler]               |       |
+| Lang-team champion                | ![Team][] [lang]                   |       |
+| Design meeting                    | ![Team][] [lang]                   |       |
+| Author call for testing blog post | *Goal point of contact, typically* |       |
```

Comment:
I'd ideally like to see stabilization for pin reborrowing only in H2 or early H1. Do you think that would fit into this goal?